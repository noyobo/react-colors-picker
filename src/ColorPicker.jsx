import React, { PropTypes }from 'react';
import ReactDOM from 'react-dom';
import Trigger from 'rc-trigger';
import ColorPickerPanel from './Panel';
import placements from './placements';

function refFn(field, component) {
  this[field] = component;
}

function prevent(e) {
  e.preventDefault();
}

export default class ColorPicker extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      color: props.color || props.defaultColor,
      alpha: props.alpha || props.defaultAlpha,
      open: false,
    };

    const events = [
      'onTriggerClick',
      'onChange',
      'onBlur',
      'getPickerElement',
      'getRootDOMNode',
      'getTriggerDOMNode',
      'onVisibleChange',
      'setOpen',
      'open',
      'close',
      'focus',
    ];

    events.forEach(e => {
      this[e] = this[e].bind(this);
    });

    this.savePickerPanelRef = refFn.bind(this, 'pickerPanelInstance');
    this.saveTriggerRef = refFn.bind(this, 'triggerInstance');
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.color) {
      this.setState({
        color: nextProps.color,
      });
    }
    if (nextProps.alpha) {
      this.setState({
        alpha: nextProps.alpha,
      });
    }
  }

  onTriggerClick() {
    this.setState({
      open: !this.state.open,
    });
  }

  onChange(colors) {
    this.setState({
      color: colors.color,
      alpha: colors.alpha,
    });
    this.props.onChange(colors);
  }

  onBlur() {
    this.setState({
      open: false,
    });
  }

  onVisibleChange(open) {
    this.setOpen(open, () => {
      if (open) {
        ReactDOM.findDOMNode(this.pickerPanelInstance).focus();
      }
    });
  }

  setOpen(open, callback) {
    const { onOpen, onClose } = this.props;
    if (this.state.open !== open) {
      this.setState({ open }, callback);
      const event = { open };
      if (open) {
        onOpen(event);
      } else {
        onClose(event);
      }
    }
  }

  getRootDOMNode() {
    return ReactDOM.findDOMNode(this);
  }

  getTriggerDOMNode() {
    return ReactDOM.findDOMNode(this.triggerInstance);
  }

  getPickerElement() {
    // const state = this.state;
    return (
        <ColorPickerPanel
          ref={ this.savePickerPanelRef }
          defaultColor={ this.state.color }
          alpha={ this.state.alpha }
          prefixCls={ `${this.props.prefixCls}-panel` }
          onChange={ this.onChange }
          onBlur={ this.onBlur }
          mode={ this.props.mode }
        />
    );
  }

  open(callback) {
    this.setOpen(true, callback);
  }

  close(callback) {
    this.setOpen(false, callback);
  }

  focus() {
    if (!this.state.open) {
      ReactDOM.findDOMNode(this).focus();
    }
  }

  render() {
    const props = this.props;
    const state = this.state;
    const classes = [props.prefixCls];
    if (state.open) {
      classes.push(`${props.prefixCls}-open`);
    }

    let trigger = props.trigger;

    if (trigger) {
      trigger = React.cloneElement(trigger, {
        ref: this.saveTriggerRef,
        unselectable: true,
        style: {
          opacity: this.state.alpha / 100,
          backgroundColor: this.state.color,
        },
        onClick: this.onTriggerClick,
        onMouseDown: prevent,
      });
    }

    const {
      prefixCls,
      placement,
      style,
      getCalendarContainer,
      align,
      animation,
      disabled,
      transitionName } = props;

    return (
      <span className={ classes.join(' ') }>
        <Trigger
          popup={ this.getPickerElement() }
          popupAlign={ align }
          builtinPlacements={ placements }
          popupPlacement={ placement }
          action={ disabled ? [] : ['click'] }
          destroyPopupOnHide
          getPopupContainer={ getCalendarContainer }
          popupStyle={ style }
          popupAnimation={ animation }
          popupTransitionName={ transitionName }
          popupVisible={ state.open }
          onPopupVisibleChange={ this.onVisibleChange }
          prefixCls={ `${prefixCls}-picker` }
        >
          { trigger }
        </Trigger>
      </span>
    );
  }
}

ColorPicker.propTypes = {
  defaultColor: PropTypes.string,
  defaultAlpha: PropTypes.number,
  // can custom
  color: PropTypes.string,
  alpha: PropTypes.number,
  onChange: PropTypes.func,
  onOpen: PropTypes.func,
  onClose: PropTypes.func,
  prefixCls: PropTypes.string.isRequired,
  trigger: PropTypes.node.isRequired,
  mode: PropTypes.oneOf(['RGB', 'HSL', 'HSB']),
  placement: PropTypes.oneOf(['topLeft', 'topRight', 'bottomLeft', 'bottomRight']),
  style: PropTypes.object,
};

ColorPicker.defaultProps = {
  defaultColor: '#F00',
  defaultAlpha: 100,
  onChange() {},
  onOpen() {},
  onClose() {},
  prefixCls: 'react-colorpicker',
  trigger: <span className="react-colorpicker-trigger"></span>,
  placement: 'topLeft',
  style: {},
};
