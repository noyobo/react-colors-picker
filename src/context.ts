import { createContext } from '@rc-component/context';
import type { Color } from './interface';

export interface ColorPickerCtxProps {
  color: Color;
  prefixCls: string;
  handleChange: (color: Color) => void;
}

const ColorPickerContext = createContext<ColorPickerCtxProps>();

export const { Provider: ColorPickerProvider } = ColorPickerContext;

export default ColorPickerContext;
