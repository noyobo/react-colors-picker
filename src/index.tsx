import ColorPicker from './ColorPicker';

export { Color } from './color';
export * from './interface';

export default ColorPicker;
