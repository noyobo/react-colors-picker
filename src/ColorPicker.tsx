import type { BuildInPlacements, TriggerProps } from '@rc-component/trigger';
import Trigger from '@rc-component/trigger';
import useMergedState from 'rc-util/lib/hooks/useMergedState';
import type { CSSProperties, FC } from 'react';
import React from 'react';
import type { ColorPickerPanelProps } from './ColorPickerPanel';
import ColorPickerPanel from './ColorPickerPanel';
import placements from './components/placements';
import { TriggerPlacement, TriggerType } from './interface';
import { ColorPickerPrefixCls } from './util';
export interface ColorPickerProps extends ColorPickerPanelProps {
  open?: boolean;
  trigger?: TriggerType;
  children: React.ReactElement;
  disabled?: boolean;
  onOpenChange?: (open: boolean) => void;
  /** Popup placement */
  placement?: TriggerPlacement;
  getPopupContainer?: (node: HTMLElement) => HTMLElement;
  classNames?: { popup?: string };
  styles?: { popup?: CSSProperties };
  builtinPlacements?: BuildInPlacements;
  arrow?: boolean;
  motion?: TriggerProps['popupMotion'];
}

const ColorPicker: FC<ColorPickerProps> = props => {
  const {
    open,
    disabled,
    trigger = 'click',
    children,
    onOpenChange,
    placement = 'bottomLeft',
    classNames,
    styles,
    prefixCls = ColorPickerPrefixCls,
    builtinPlacements = placements,
    motion,
    ...resetProps
  } = props;

  const [openValue, setOpenValue] = useMergedState(false, {
    value: open,
    postState: openData => !disabled && openData,
    onChange: onOpenChange,
  });

  return (
    <Trigger
      action={[trigger]}
      popupVisible={openValue}
      popup={<ColorPickerPanel {...props} />}
      popupPlacement={placement}
      onPopupVisibleChange={setOpenValue}
      popupClassName={classNames?.popup}
      popupStyle={styles?.popup}
      builtinPlacements={builtinPlacements}
      popupMotion={motion}
      prefixCls={prefixCls}
      {...resetProps}
    >
      {children}
    </Trigger>
  );
};

export default ColorPicker;
