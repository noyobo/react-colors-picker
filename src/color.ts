import type { HSVA, Numberify } from '@ctrl/tinycolor';
import { TinyColor } from '@ctrl/tinycolor';
import type { ColorGenInput, HSBA } from './interface';
import { convertHsb2Hsv, getRoundNumber } from './util';
export class Color extends TinyColor {
  constructor(color: ColorGenInput) {
    super(convertHsb2Hsv(color));
  }

  toHsbString() {
    const hsb = this.toHsb();
    const saturation = getRoundNumber(hsb.s * 100);
    const lightness = getRoundNumber(hsb.b * 100);
    const hue = getRoundNumber(hsb.h);
    const alpha = hsb.a;
    const hsbString = `hsb(${hue}, ${saturation}%, ${lightness}%)`;
    const hsbaString = `hsba(${hue}, ${saturation}%, ${lightness}%, ${alpha.toFixed(
      alpha === 0 ? 0 : 2,
    )})`;
    return alpha === 1 ? hsbString : hsbaString;
  }

  toHsb(): Numberify<HSBA> {
    let hsv = this.toHsv();
    if (typeof this.originalInput === 'object' && this.originalInput) {
      hsv = this.originalInput as Numberify<HSVA>;
    }
    const { v, ...resets } = hsv;
    return {
      ...resets,
      b: hsv.v,
    };
  }
}
