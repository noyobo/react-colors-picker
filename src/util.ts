import type { ColorInput } from '@ctrl/tinycolor';
import { Color } from './color';
import type {
  ColorGenInput,
  HSB,
  HsbaColorType,
  TransformOffset,
} from './interface';

export const ColorPickerPrefixCls = 'rc-color';

export const getRoundNumber = (value: number) => Math.round(Number(value || 0));

export const convertHsb2Hsv = (color: ColorGenInput): ColorInput => {
  if (
    typeof color === 'object' &&
    Reflect.has(color, 'h') &&
    Reflect.has(color, 'b')
  ) {
    const { b, ...resets } = color as HSB;
    return {
      ...resets,
      v: b,
    };
  }
  if (typeof color === 'string' && /hsb/.test(color)) {
    return color.replace(/hsb/, 'hsv');
  }
  return color as ColorInput;
};

export const generateColor = (color: ColorGenInput): Color => {
  if (color instanceof Color) {
    return color;
  }
  return new Color(color);
};

export const defaultColor = generateColor('#1677ff');

export const calculateColor = (props: {
  offset: TransformOffset;
  containerRef: React.RefObject<HTMLDivElement>;
  targetRef: React.RefObject<HTMLDivElement>;
  color?: Color;
  type?: HsbaColorType;
}): Color => {
  const { offset, targetRef, containerRef, color, type } = props;
  const { width, height } = containerRef.current.getBoundingClientRect();
  const { width: targetWidth, height: targetHeight } =
    targetRef.current.getBoundingClientRect();
  const centerOffsetX = targetWidth / 2;
  const centerOffsetY = targetHeight / 2;
  const saturation = (offset.x + centerOffsetX) / width;
  const bright = 1 - (offset.y + centerOffsetY) / height;
  const hsb = color.toHsb();
  const alphaOffset = saturation;
  const hueOffset = ((offset.x + centerOffsetX) / width) * 360;

  if (type) {
    switch (type) {
      case 'hue':
        return generateColor({
          ...hsb,
          h: hueOffset <= 0 ? 0 : hueOffset,
        });
      case 'alpha':
        return generateColor({
          ...hsb,
          a: alphaOffset <= 0 ? 0 : alphaOffset,
        });
    }
  }

  return generateColor({
    h: hsb.h,
    s: saturation <= 0 ? 0 : saturation,
    b: bright >= 1 ? 1 : bright,
    a: hsb.a,
  });
};

export const calculateOffset = (
  containerRef: React.RefObject<HTMLDivElement>,
  targetRef: React.RefObject<HTMLDivElement>,
  color?: Color,
  type?: HsbaColorType,
): TransformOffset => {
  const { width, height } = containerRef.current.getBoundingClientRect();
  const { width: targetWidth, height: targetHeight } =
    targetRef.current.getBoundingClientRect();
  const centerOffsetX = targetWidth / 2;
  const centerOffsetY = targetHeight / 2;
  const hsb = color.toHsb();

  // Exclusion of boundary cases
  if (
    (targetWidth === 0 && targetHeight === 0) ||
    targetWidth !== targetHeight
  ) {
    return;
  }

  if (type) {
    switch (type) {
      case 'hue':
        return {
          x: (hsb.h / 360) * width - centerOffsetX,
          y: -centerOffsetY / 3,
        };
      case 'alpha':
        return {
          x: (hsb.a / 1) * width - centerOffsetX,
          y: -centerOffsetY / 3,
        };
    }
  }
  return {
    x: hsb.s * width - centerOffsetX,
    y: (1 - hsb.b) * height - centerOffsetY,
  };
};
