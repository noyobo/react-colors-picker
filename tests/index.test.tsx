import { act, createEvent, fireEvent, render } from '@testing-library/react';
import { spyElementPrototypes } from 'rc-util/lib/test/domHook';
import React, { useState } from 'react';
import ColorPicker from '../src';
import { defaultColor } from '../src/util';

vi.mock('@rc-component/trigger', async () => {
  return await import('@rc-component/trigger/lib/mock');
});

export async function waitFakeTimer(advanceTime = 1000, times = 20) {
  for (let i = 0; i < times; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    await act(async () => {
      await Promise.resolve();

      if (advanceTime > 0) {
        vi.advanceTimersByTime(advanceTime);
      } else {
        vi.runAllTimers();
      }
    });
  }
}

function doMouseMove(
  container,
  start,
  end,
  element = 'rc-color-picker-handler',
) {
  const mouseDown = createEvent.mouseDown(
    container.getElementsByClassName(element)[0],
    {
      pageX: start,
      pageY: start,
    },
  );
  fireEvent(container.getElementsByClassName(element)[0], mouseDown);

  // Drag
  const mouseMove: any = new Event('mousemove');
  mouseMove.pageX = end;
  mouseMove.pageY = end;

  fireEvent(document, mouseMove);

  const mouseUp = createEvent.mouseUp(document);
  fireEvent(document, mouseUp);
}

function doTouchMove(
  container,
  start,
  end,
  element = 'rc-color-picker-handler',
) {
  const touchStart: any = createEvent.touchStart(
    container.getElementsByClassName(element)[0],
    {
      touches: [{}],
    },
  );
  touchStart.touches[0].pageX = start;
  fireEvent(container.getElementsByClassName(element)[0], touchStart);

  // Drag
  const touchMove: any = createEvent.touchMove(document, {
    touches: [{}],
  });
  touchMove.touches[0].pageX = end;
  fireEvent(document, touchMove);
}

describe('ColorPicker', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });
  afterEach(() => {
    vi.useRealTimers();
  });
  it('Should component render correct', () => {
    const { container } = render(
      <ColorPicker open defaultValue={defaultColor}>
        <div>Color Picker</div>
      </ColorPicker>,
    );
    expect(container).toMatchSnapshot();
    expect(container.querySelector('.rc-color-picker-panel')).toBeTruthy();
  });

  it('Should trigger work', () => {
    const { container: clickContainer } = render(
      <ColorPicker trigger="click">
        <div className="trigger">Color Picker</div>
      </ColorPicker>,
    );
    // click
    expect(clickContainer.querySelector('.rc-color-picker-panel')).toBeFalsy();
    fireEvent.click(clickContainer.querySelector('.trigger'));
    expect(clickContainer.querySelector('.rc-color-picker-panel')).toBeTruthy();
    fireEvent.click(clickContainer.querySelector('.trigger'));
    expect(document.body.querySelector('.rc-color-picker-hidden')).toBeTruthy();

    const { container: hoverContainer } = render(
      <ColorPicker trigger="hover">
        <div className="trigger">Color Picker</div>
      </ColorPicker>,
    );

    // hover
    expect(hoverContainer.querySelector('.rc-color-picker-panel')).toBeFalsy();
    fireEvent.mouseEnter(hoverContainer.querySelector('.trigger'));
    expect(hoverContainer.querySelector('.rc-color-picker-panel')).toBeTruthy();
    fireEvent.mouseLeave(hoverContainer.querySelector('.trigger'));
    expect(document.body.querySelector('.rc-color-picker-hidden')).toBeTruthy();
  });

  it('Should pick color work by mouse', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <ColorPicker open value={value} onChange={value => setValue(value)}>
          <>
            <div className="pick-color">{value.toHsbString()}</div>
            <div>{value.toHexString()}</div>
            <div>{value.toHex8String()}</div>
          </>
        </ColorPicker>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );

    doMouseMove(container, 0, 999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(360, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      9999,
      0,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsba(360, 100%, 0%, 0)',
    );
  });

  it('Should no control pick color work', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <ColorPicker open onChange={value => setValue(value)}>
          <>
            <div className="pick-color">{value.toHsbString()}</div>
            <div>{value.toHexString()}</div>
            <div>{value.toHex8String()}</div>
          </>
        </ColorPicker>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );

    doMouseMove(container, 0, 9999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(360, 100%, 0%)',
    );

    doMouseMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      9999,
      0,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsba(360, 100%, 0%, 0)',
    );
  });

  it('Should pick color work by touch', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <ColorPicker open value={value} onChange={value => setValue(value)}>
          <div className="pick-color">{value.toHsbString()}</div>
        </ColorPicker>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );

    doTouchMove(container, 0, 9999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 100%, 0%)',
    );

    doTouchMove(
      container.querySelector('.rc-color-picker-slider-hue'),
      0,
      9999,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(360, 100%, 0%)',
    );

    doTouchMove(
      container.querySelector('.rc-color-picker-slider-alpha'),
      9999,
      0,
    );
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsba(360, 100%, 0%, 0)',
    );
  });

  it('Should custom panel work', () => {
    const { container } = render(
      <ColorPicker
        open
        panelRender={panel => <div className="custom-panel">{panel}</div>}
      >
        <div>Color Picker</div>
      </ColorPicker>,
    );
    expect(container.querySelector('.custom-panel')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('Should prefixCls work', () => {
    const { container } = render(
      <ColorPicker open prefixCls="test-prefixCls">
        <div>Color Picker</div>
      </ColorPicker>,
    );
    expect(container).toMatchSnapshot();
  });

  it('Should drag boundary cases work', () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 0,
        height: 0,
      }),
    });
    const App = () => {
      const [value, setValue] = useState(defaultColor);
      return (
        <ColorPicker open value={value} onChange={value => setValue(value)}>
          <div className="pick-color">{value.toHsbString()}</div>
        </ColorPicker>
      );
    };
    const { container } = render(<App />);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );
    doMouseMove(container, 0, 9999);
    expect(container.querySelector('.pick-color').innerHTML).toBe(
      'hsb(215, 91%, 100%)',
    );
  });

  it('Should hsb string work', () => {
    const App = () => (
      <ColorPicker open value={'hsb(215, 91%, 100%)'}>
        <div>Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should rgb string work', () => {
    const App = () => (
      <ColorPicker open value={'rgb(23, 120, 255)'}>
        <div>Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should hex string work', () => {
    const App = () => (
      <ColorPicker open value={'#1778ff'}>
        <div>Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should hsb obj work', () => {
    const App = () => (
      <ColorPicker open value={{ h: 215, s: 0.91, b: 1 }}>
        <div>Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should rgb obj work', () => {
    const App = () => (
      <ColorPicker open value={{ r: 23, g: 120, b: 255 }}>
        <div>Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker-handler').getAttribute('style'),
    ).toEqual('background-color: rgb(23, 120, 255);');
  });

  it('Should disabled work', () => {
    const App = () => (
      <ColorPicker disabled>
        <div className="trigger">Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);

    // click
    fireEvent.click(container.querySelector('.trigger'));
    expect(container.querySelector('.rc-color-picker-panel')).toBeFalsy();
  });

  it('Should component open work', async () => {
    const handleOpenChange = vi.fn();
    const App = () => {
      const [open, setOpen] = useState(false);
      return (
        <ColorPicker open={open} onOpenChange={handleOpenChange}>
          <div className="trigger" onClick={() => setOpen(!open)}>
            Color Picker
          </div>
        </ColorPicker>
      );
    };
    const { container } = render(<App />);

    fireEvent.click(container.querySelector('.trigger'));
    await waitFakeTimer();
    expect(container.querySelector('.rc-color-picker-panel')).toBeTruthy();
    expect(handleOpenChange).toHaveBeenCalledWith(true, false);

    fireEvent.click(container.querySelector('.trigger'));
    await waitFakeTimer();
    expect(document.body.querySelector('.rc-color-picker-hidden')).toBeTruthy();
    expect(handleOpenChange).toHaveBeenLastCalledWith(false, true);
  });

  it('Should styles work', async () => {
    const App = () => (
      <ColorPicker
        styles={{
          popup: {
            width: 500,
          },
        }}
        open
      >
        <div className="trigger">Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    expect(
      container.querySelector('.rc-color-picker').getAttribute('style'),
    ).toContain('width: 500px;');
  });

  it('Should fix open misuse work', async () => {
    spyElementPrototypes(HTMLElement, {
      getBoundingClientRect: () => ({
        x: 0,
        y: 100,
        width: 100,
        height: 100,
      }),
    });
    const App = () => (
      <ColorPicker
        styles={{
          popup: {
            width: 500,
          },
        }}
      >
        <div className="trigger">Color Picker</div>
      </ColorPicker>
    );
    const { container } = render(<App />);
    fireEvent.click(container.querySelector('.trigger'));
    await waitFakeTimer();
    expect(container.querySelector('.rc-color-picker-panel')).toBeTruthy();
    doMouseMove(container, 0, 9999);
    doTouchMove(container, 0, 9999);
    expect(document.body.querySelector('.rc-color-picker-hidden')).toBeFalsy();

    fireEvent.click(container.querySelector('.trigger'));
    await waitFakeTimer();
    fireEvent.click(document);
    await waitFakeTimer();
    expect(document.body.querySelector('.rc-color-picker-hidden')).toBeTruthy();
  });
});
