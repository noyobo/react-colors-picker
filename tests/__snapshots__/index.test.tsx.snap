// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker Should component render correct 1`] = `
<div>
  <div>
    Color Picker
  </div>
  <div
    class="rc-color-picker rc-color-picker-placement-bottomLeft"
    style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
  >
    <div
      class="rc-color-picker-panel"
    >
      <div
        class="rc-color-picker-select"
      >
        <div
          class="rc-color-picker-palette"
          style="position: relative;"
        >
          <div
            style="position: absolute; left: 0px; top: 0px; z-index: 1;"
          >
            <div
              class="rc-color-picker-handler"
              style="background-color: rgb(22, 119, 255);"
            />
          </div>
          <div
            class="rc-color-picker-saturation"
            style="background-color: rgb(0, 0, 0);"
          />
        </div>
      </div>
      <div
        class="rc-color-picker-slider-container"
      >
        <div
          class="rc-color-picker-slider-group"
        >
          <div
            class="rc-color-picker-slider rc-color-picker-slider-hue"
          >
            <div
              class="rc-color-picker-palette"
              style="position: relative;"
            >
              <div
                style="position: absolute; left: 0px; top: 0px; z-index: 1;"
              >
                <div
                  class="rc-color-picker-handler rc-color-picker-handler-sm"
                  style="background-color: rgb(0, 0, 0);"
                />
              </div>
              <div
                class="rc-color-picker-gradient"
                style="position: absolute; inset: 0;"
              />
            </div>
          </div>
          <div
            class="rc-color-picker-slider rc-color-picker-slider-alpha"
          >
            <div
              class="rc-color-picker-palette"
              style="position: relative;"
            >
              <div
                style="position: absolute; left: 0px; top: 0px; z-index: 1;"
              >
                <div
                  class="rc-color-picker-handler rc-color-picker-handler-sm"
                  style="background-color: rgb(22, 119, 255);"
                />
              </div>
              <div
                class="rc-color-picker-gradient"
                style="position: absolute; inset: 0;"
              />
            </div>
          </div>
        </div>
        <div
          class="rc-color-picker-color-block"
        >
          <div
            class="rc-color-picker-color-block-inner"
            style="background: rgb(22, 119, 255);"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPicker Should custom panel work 1`] = `
<div>
  <div>
    Color Picker
  </div>
  <div
    class="rc-color-picker rc-color-picker-placement-bottomLeft"
    style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
  >
    <div
      class="rc-color-picker-panel"
    >
      <div
        class="custom-panel"
      >
        <div
          class="rc-color-picker-select"
        >
          <div
            class="rc-color-picker-palette"
            style="position: relative;"
          >
            <div
              style="position: absolute; left: 41.372549019607845px; top: -50px; z-index: 1;"
            >
              <div
                class="rc-color-picker-handler"
                style="background-color: rgb(22, 119, 255);"
              />
            </div>
            <div
              class="rc-color-picker-saturation"
              style="background-color: rgb(0, 0, 0);"
            />
          </div>
        </div>
        <div
          class="rc-color-picker-slider-container"
        >
          <div
            class="rc-color-picker-slider-group"
          >
            <div
              class="rc-color-picker-slider rc-color-picker-slider-hue"
            >
              <div
                class="rc-color-picker-palette"
                style="position: relative;"
              >
                <div
                  style="position: absolute; left: 9.728183118741065px; top: -16.666666666666668px; z-index: 1;"
                >
                  <div
                    class="rc-color-picker-handler rc-color-picker-handler-sm"
                    style="background-color: rgb(0, 0, 0);"
                  />
                </div>
                <div
                  class="rc-color-picker-gradient"
                  style="position: absolute; inset: 0;"
                />
              </div>
            </div>
            <div
              class="rc-color-picker-slider rc-color-picker-slider-alpha"
            >
              <div
                class="rc-color-picker-palette"
                style="position: relative;"
              >
                <div
                  style="position: absolute; left: 50px; top: -16.666666666666668px; z-index: 1;"
                >
                  <div
                    class="rc-color-picker-handler rc-color-picker-handler-sm"
                    style="background-color: rgb(22, 119, 255);"
                  />
                </div>
                <div
                  class="rc-color-picker-gradient"
                  style="position: absolute; inset: 0;"
                />
              </div>
            </div>
          </div>
          <div
            class="rc-color-picker-color-block"
          >
            <div
              class="rc-color-picker-color-block-inner"
              style="background: rgb(22, 119, 255);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`ColorPicker Should prefixCls work 1`] = `
<div>
  <div>
    Color Picker
  </div>
  <div
    class="test-prefixCls test-prefixCls-placement-bottomLeft"
    style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
  >
    <div
      class="test-prefixCls-panel"
    >
      <div
        class="test-prefixCls-select"
      >
        <div
          class="test-prefixCls-palette"
          style="position: relative;"
        >
          <div
            style="position: absolute; left: 41.372549019607845px; top: -50px; z-index: 1;"
          >
            <div
              class="test-prefixCls-handler"
              style="background-color: rgb(22, 119, 255);"
            />
          </div>
          <div
            class="test-prefixCls-saturation"
            style="background-color: rgb(0, 0, 0);"
          />
        </div>
      </div>
      <div
        class="test-prefixCls-slider-container"
      >
        <div
          class="test-prefixCls-slider-group"
        >
          <div
            class="test-prefixCls-slider test-prefixCls-slider-hue"
          >
            <div
              class="test-prefixCls-palette"
              style="position: relative;"
            >
              <div
                style="position: absolute; left: 9.728183118741065px; top: -16.666666666666668px; z-index: 1;"
              >
                <div
                  class="test-prefixCls-handler test-prefixCls-handler-sm"
                  style="background-color: rgb(0, 0, 0);"
                />
              </div>
              <div
                class="test-prefixCls-gradient"
                style="position: absolute; inset: 0;"
              />
            </div>
          </div>
          <div
            class="test-prefixCls-slider test-prefixCls-slider-alpha"
          >
            <div
              class="test-prefixCls-palette"
              style="position: relative;"
            >
              <div
                style="position: absolute; left: 50px; top: -16.666666666666668px; z-index: 1;"
              >
                <div
                  class="test-prefixCls-handler test-prefixCls-handler-sm"
                  style="background-color: rgb(22, 119, 255);"
                />
              </div>
              <div
                class="test-prefixCls-gradient"
                style="position: absolute; inset: 0;"
              />
            </div>
          </div>
        </div>
        <div
          class="test-prefixCls-color-block"
        >
          <div
            class="test-prefixCls-color-block-inner"
            style="background: rgb(22, 119, 255);"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
